bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringTable.java

```diff
@@ -0,0 +1,1052 @@
+
+  private final SecondaryWriteErrorConsumer secondaryWriteErrorConsumer;
+  private final MirroringTracer mirroringTracer;
+
+  private final ReadSampler readSampler;
+  private final boolean performWritesConcurrently;
+  /**
+   * @param executorService ExecutorService is used to perform operations on secondaryTable and
+   *     verification tasks.
+   * @param mismatchDetector Detects mismatches in results from operations preformed on both
+   *     databases.
+   * @param secondaryWriteErrorConsumer Consumer secondary write errors.
+   */
+  public MirroringTable(
+      Table primaryTable,
+      Table secondaryTable,
+      ExecutorService executorService,
+      MismatchDetector mismatchDetector,
+      FlowController flowController,
+      SecondaryWriteErrorConsumer secondaryWriteErrorConsumer,
+      ReadSampler readSampler,
+      boolean performWritesConcurrently,
+      MirroringTracer mirroringTracer) {
+    this.primaryTable = primaryTable;
+    this.secondaryTable = secondaryTable;
+    this.verificationContinuationFactory = new VerificationContinuationFactory(mismatchDetector);
+    this.readSampler = readSampler;
+    this.secondaryAsyncWrapper =
+        new AsyncTableWrapper(
+            this.secondaryTable,
+            MoreExecutors.listeningDecorator(executorService),
+            mirroringTracer);
+    this.flowController = flowController;
+    this.referenceCounter = new ListenableReferenceCounter();
+    this.referenceCounter.holdReferenceUntilClosing(this.secondaryAsyncWrapper);
+    this.secondaryWriteErrorConsumer = secondaryWriteErrorConsumer;
+    this.performWritesConcurrently = performWritesConcurrently;
+    this.mirroringTracer = mirroringTracer;
+  }
+
+  @Override
+  public TableName getName() {
+    return this.primaryTable.getName();
+  }
+
+  @Override
+  public Configuration getConfiguration() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public HTableDescriptor getTableDescriptor() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean exists(final Get get) throws IOException {
+    try (Scope scope = this.mirroringTracer.spanFactory.operationScope(HBaseOperation.EXISTS)) {
+      Log.trace("[%s] exists(get=%s)", this.getName(), get);
+
+      boolean result =
+          this.mirroringTracer.spanFactory.wrapPrimaryOperation(
+              new CallableThrowingIOException<Boolean>() {
+                @Override
+                public Boolean call() throws IOException {
+                  return MirroringTable.this.primaryTable.exists(get);
+                }
+              },
+              HBaseOperation.EXISTS);
+
+      if (this.readSampler.shouldNextReadOperationBeSampled()) {
+        scheduleSequentialReadOperationWithVerification(
+            new RequestResourcesDescription(result),
+            this.secondaryAsyncWrapper.exists(get),
+            this.verificationContinuationFactory.exists(get, result));
+      }
```

Comment:
I think it would be nicer to encapsulate the whole lot:

verifyReadOperation(
   verification.exists(get, result),
   () -> secondaryAsyncWrapper.exists(get)
)

Then verifyReadOperation can check readSampler and build the description and the verification callback

Reply:
Done (but created helper method also takes resource description as a parameter).